src/utils/material.js

```diff
@@ -97,24 +98,55 @@ module.exports.updateMapMaterialFromData = function (materialName, dataName, sha
 
   // If the new material src is already a texture, just use it.
   if (src instanceof THREE.Texture) { setMap(src); } else {
-    // Load texture for the new material src.
+    // Load texture source for the new material src.
     // (And check if we should still use it once available in callback.)
-    el.sceneEl.systems.material.loadTexture(src,
-      {src: src, repeat: data.repeat, offset: data.offset, npot: data.npot, anisotropy: data.anisotropy},
-      checkSetMap);
+    el.sceneEl.systems.material.loadTextureSource(src, updateTexture);
   }
 
-  function checkSetMap (texture) {
+  function updateTexture (source) {
     // If the source has been changed, don't use loaded texture.
     if (shader.materialSrcs[materialName] !== src) { return; }
+
+    var texture = material[materialName];
+
+    // Handle removal or texture type change
+    if (texture && (source === null || !isCompatibleTexture(texture, source))) {
+      texture = null;
+    }
+
+    // Create texture if needed
+    if (!texture && source) {
+      texture = createCompatibleTexture(source);
+    }
+
+    // Update texture source and properties
+    if (texture) {
+      if (texture.source !== source) {
+        texture.source = source;
+        texture.needsUpdate = true;
+      }
+      if (COLOR_MAPS.has(materialName)) {
+        rendererSystem.applyColorCorrection(texture);
+      }
+      setTextureProperties(texture, data);
+    }
+
+    // Set map property on the material
     setMap(texture);
   }
 
   function setMap (texture) {
-    material[materialName] = texture;
-    if (texture && COLOR_MAPS.has(materialName)) {
-      rendererSystem.applyColorCorrection(texture);
+    // Nothing to do if texture is the same
+    if (material[materialName] === texture) {
+      return;
+    }
+
+    // Dispose old texture if present
+    if (material[materialName]) {
+      material[materialName].dispose();
```

Comment:
Before this PR, the texture wouldn't be disposed when switching, effectively 'leaking' them. This did have the benefit of keeping it on the GPU in case the texture would be used later on. For example when a user repeatedly switches between two textures.

With this change, upon disposal the texture will be freed from the GPU (unless it's still being used elsewhere). So when reintroducing the texture in the scene, an upload takes place. The repeatedly switching would then incur a texture upload every time it switches.

The new behaviour is definitely more correct since A-Frame has no other way to track the (intended) lifecycle of a texture. But we might need to provide users with ways to keep textures 'alive'. Do note, however, that it's about the specific "texture source + texture properties". One such way might be to keep materials alive, see #5457